Sources/CodeGen/CXX/CodeFormatter.swift

```diff
@@ -0,0 +1,88 @@
+/// Class responsible for the formatiing of the CXX code, as it is generated.
+///
+/// The callers will tell this class what tokens they want to write to the output CXX code,
+/// and this will ensure that proper formatting will be added.
+///
+/// This class will have ownership of the code that is being written.
+struct CodeFormatter: TextOutputStream {
+
+  /// The amount of indentation we apply to the code.
+  private var indentSize: Int
+
+  /// The code that this object is writing.
+  var code: String = ""
+
+  /// The current level of indentation.
+  private var curIndent: Int = 0
+  /// True if the next text written should be indented.
+  private var shouldIntent: Bool = false
+
+  /// Initializes this object with the right formatting parameters.
+  init(indentSize: Int = 2) {
+    self.indentSize = indentSize
+  }
+
+  /// Adds new text to the output code.
+  mutating func write(_ text: String) {
+    checkAddIndent()
+    code.write(text)
+  }
+
+  /// Adds new text to the output code, with a new line at the end
+  mutating func writeLn(_ text: String) {
```

Comment:
```suggestion
  mutating func writeLine(_ text: String) {
```

I would stay away from abbreviations when possible. I think we can afford the two extra characters here.

Reply:
There are multiple programming languages with `writeLn` function. But, yes, we are just adding an utility...